core/src/main/java/fr/sncf/osrd/train/RollingStock.java

```diff
@@ -207,17 +206,14 @@ private TractiveEffortPoint[] getTractiveEffortCurve(String profile, Comfort com
      */
     public double getMaxEffort(double speed, String profile, Comfort comfort) {
         final var tractiveEffortCurve = getTractiveEffortCurve(profile, comfort);
-        double previousEffort = 0.0;
-        double previousSpeed = 0.0;
-        for (var dataPoint : tractiveEffortCurve) {
-            if (previousSpeed <= Math.abs(speed) && Math.abs(speed) < dataPoint.speed) {
-                var coeff = (previousEffort - dataPoint.maxEffort) / (previousSpeed - dataPoint.speed);
-                return previousEffort + coeff * (Math.abs(speed) - previousSpeed);
-            }
-            previousEffort = dataPoint.maxEffort;
-            previousSpeed = dataPoint.speed;
-        }
-        return previousEffort;
+        Arrays.sort(tractiveEffortCurve);
+        double[] tractiveEffortCurveSpeeds =
+                Arrays.stream(tractiveEffortCurve).mapToDouble(tractiveEffortPoint -> speed).toArray();
+        int index = Arrays.binarySearch(tractiveEffortCurveSpeeds, speed);
+        TractiveEffortPoint previousPoint = tractiveEffortCurve[index];
```

Comment:
Have you tested your code? I don't think it works: binarySearch returns a negative value when a point is between two others

Reply:
you're right. I will correct it